### DISCUSSION TITLE
Allow to set registry mirror

### DISCUSSION THREAD

**Discussion by mokazemi** [nextcloud/all-in-one#5268] [Ideas] [answered]
Hi. The server that I'm using for Nextcloud AIO is located in Iran and have some problems in accessing to auth.docker.io. I usually set a registry mirror for my docker like this:
```
{
  "registry-mirrors": ["https://registry.docker.ir"]
}
```
But the issue is, the `nextcloud/all-in-one`  internally tires to access to the docker.io to check for updates. Now I can't even open the Nextcloud AIO Interface page, and I see some errors on the logs of the container:

```
[Sun Sep 15 08:27:09.307602 2024] [core:notice] [pid 188:tid 188] AH00094: Command line: 'httpd -D FOREGROUND'
NOTICE: PHP message: Could not get digest of container nextcloud/all-in-one:latest cURL error 28: Failed to connect to auth.docker.io port 443 after 300002 ms: Timeout was reached (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for https://auth.docker.io/token?service=registry.docker.io&scope=repository:nextcloud/all-in-one:pull
NOTICE: PHP message: Could not get digest of container nextcloud/aio-apache:latest cURL error 28: Failed to connect to auth.docker.io port 443 after 300000 ms: Timeout was reached (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for https://auth.docker.io/token?service=registry.docker.io&scope=repository:nextcloud/aio-apache:pull
```

How should I solve the issue? Could there be an option to use system registry mirrors instead of using the default one or a env var for it? Or anything else I can do?

**Comment by szaimen** [comment]
Hi, thanks for your report. In general we might be able to retrieve this value via docker info. However I would need to know where to reach the authentication domain as this is apparently the one that is not reachable:

> Timeout was reached (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for https://auth.docker.io

**Reply by szaimen** [reply]
Is there maybe also a mirror for that?

**Reply by mokazemi** [reply]
Oh, sorry I didn't see your reply.
Could you please explain more on this? What info should I provide?

**Reply by szaimen** [reply]
See https://github.com/nextcloud/all-in-one/discussions/5268#discussioncomment-11369639

**Comment by houming818** [comment]
The problem is the same as in .cn

**Reply by szaimen** [reply]
I would need to know where to reach the authentication domain as this is apparently the one that is not reachable:

> Timeout was reached (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for https://auth.docker.io/

Is there maybe also a mirror for that?

**Reply by aojie654** [reply]
@szaimen Can we config it to self-buildup registry mirror in anywhere please?

**Reply by szaimen** [reply]
Can you please answer my question above?

**Reply by aojie654** [reply]
Hi Simon,

Is it possible to use a config file or variable to set the registry mirror for some domains? e.g.

``` ini
[MIRRORS]
auth.docker.io = mirror-auth.example.com
hub.docker.com = mirror-hub.example.com
```

**Reply by szaimen** [reply]
Currently it is not but I was wondering if there is a value in docker itself where you can modify auth.docker.io and hub.docker.com in the daemon.json of docker that we could retreive maybe via `sudo docker info`?

**Reply by aojie654** [reply]
Looks there is too many issues from some region with network blocking, so I set up Nextcloud manually with php-8.3 and Nginx. 🤔

**Comment by invis-z** [comment]
> I would need to know where to reach the authentication domain as this is apparently the one that is not reachable:
> 
> > Timeout was reached (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for https://auth.docker.io/
> 
> Is there maybe also a mirror for that?

@szaimen Based on my understanding, the `auth.docker.io` domain is not in any part of the code, instead it comes from the registry server 
```
$ curl -iL index.docker.io/v2
HTTP/1.1 301 Moved Permanently
content-length: 0
location: https://index.docker.io/v2

HTTP/1.1 301 Moved Permanently
content-type: text/html; charset=utf-8
docker-distribution-api-version: registry/2.0
location: /v2/
date: Thu, 16 Jan 2025 07:18:13 GMT
content-length: 39
strict-transport-security: max-age=31536000

HTTP/1.1 401 Unauthorized
content-type: application/json
docker-distribution-api-version: registry/2.0
www-authenticate: Bearer realm="https://auth.docker.io/token",service="registry.docker.io"
date: Thu, 16 Jan 2025 07:18:13 GMT
content-length: 87
strict-transport-security: max-age=31536000

{"errors":[{"code":"UNAUTHORIZED","message":"authentication required","detail":null}]}
```

In the last response there is a `www-authenticate` header and it gives the domain. In theory, the mirror will return its own domain and url that would be different for each server. 

PS. I referenced watchtower's code to figure this out as I was also facing this problem with it. [Link](https://github.com/containrrr/watchtower/blob/76f9cea516593fabb8ca91ff13de55caa6aa0a8b/pkg/registry/auth/auth.go)

**Reply by invis-z** [reply]
Quickly spun up a container for a local registry and here is what I got:
```
$ curl -iL localhost:5000/v2
HTTP/1.1 301 Moved Permanently
Content-Type: text/html; charset=utf-8
Docker-Distribution-Api-Version: registry/2.0
Location: /v2/
Date: Thu, 16 Jan 2025 07:49:14 GMT
Content-Length: 39

HTTP/1.1 401 Unauthorized
Content-Type: application/json; charset=utf-8
Docker-Distribution-Api-Version: registry/2.0
Www-Authenticate: Basic realm="Registry Realm"
X-Content-Type-Options: nosniff
Date: Thu, 16 Jan 2025 07:49:14 GMT
Content-Length: 87

{"errors":[{"code":"UNAUTHORIZED","message":"authentication required","detail":null}]}
```
In this case, there will not be another domain for the authentication and it will be done through http basic auth.

Here basic auth is what I set up, it might possible to do `Bearer` but the set up will be more complicated. In that case it will return a similar header as the official registry but with the server's own domain. 

If the mirror didn't set up auth, it will just return a 200:
```
$ curl -iL https://my-local-mirror.example/v2
HTTP/2 301 
content-type: text/html; charset=utf-8
docker-distribution-api-version: registry/2.0
location: /v2/
content-length: 39
date: Thu, 16 Jan 2025 07:53:46 GMT

HTTP/2 200 
content-type: application/json; charset=utf-8
docker-distribution-api-version: registry/2.0
x-content-type-options: nosniff
content-length: 2
date: Thu, 16 Jan 2025 07:53:46 GMT

{}
```

**Reply by invis-z** [reply]
Also, by the way, **containrrr/watchtower** seems to be inactive. And the mirror will not work with it: containrrr/watchtower#1346. A fork or a switch to other projects might be needed.

**Comment by szaimen** [comment]
Hi, thanks for all the input. After thinking a bit more about this issue, the best solution from my POV is to add a check that blocks the installation if access to auth.docker.io is not possible: https://github.com/nextcloud/all-in-one/pull/5888

However it will also point everyone at https://github.com/nextcloud/all-in-one/tree/main/manual-install which allows to install the project in such environments.

**Reply by invis-z** [reply]
I would argue it would be better to change the way auth is handled. The only place the domain is hardcoded is at https://github.com/nextcloud/all-in-one/blob/bdff564e2ed90a84dddb8938ead7f5cd017ed7d9/php/src/Docker/DockerHubManager.php#L30 and in my opinion the auth logic should be revised to the more standard way of querying the registry and follow its requirement, like https://github.com/containrrr/watchtower/blob/76f9cea516593fabb8ca91ff13de55caa6aa0a8b/pkg/registry/auth/auth.go even if registry mirror remain unsupported.

**Reply by invis-z** [reply]
If you could hold off merging #5888 for a little bit, I am currently having some free time and can try to ctrl-c ctrl-v something together. Since currently it doesn't do too much harm to not merging it.

**Reply by szaimen** [reply]
This can be done in a follow-up and then remove the check again. However since watchtower is an integral part of AIO and https://github.com/containrrr/watchtower/issues/1346 is still open, I fear it doesnt not make much sense to implement this at all on the AIO side...

**Reply by invis-z** [reply]
https://github.com/containrrr/watchtower/issues/1346 can be easily bypassed by adding the mirror link before the repository, like change from `nextcloud/aio-mastercontainer` to `https://my-doeker-mirror/nextcloud/aio-mastercontainer`

**Reply by invis-z** [reply]
This also can be done for docker itself, so if add a `CUSTOM_REGISTRY` env, we can bypass the whole docker mirror setting thing completely

**Reply by invis-z** [reply]
Sorry, I did not think this through, after some digging, it seems that watchtower is only used to update the mastercontainer, so when user installing, they can just use `my-mirror/nextcloud/all-in-one` instead of `nextcloud/all-in-one`, and if mirror is supported on the AIO side, everything should just work.

**Reply by szaimen** [reply]
So what about auth.docker.io? Is this not needed in the case of a mirror?

**Reply by szaimen** [reply]
We need it in the normal installation to retreive a token so that we can check for image updates without downloading them

**Reply by invis-z** [reply]
The auth.docker.io should never been hardcoded, it should be queried from `index.docker.io/v2` or `registry-1.docker.io/v2` from the www-authenticate header of the response.

```
$ curl -iL registry-1.docker.io/v2
HTTP/1.1 301 Moved Permanently
content-length: 0
location: https://registry-1.docker.io/v2

HTTP/1.1 301 Moved Permanently
content-type: text/html; charset=utf-8
docker-distribution-api-version: registry/2.0
location: /v2/
date: Thu, 16 Jan 2025 10:51:12 GMT
content-length: 39
strict-transport-security: max-age=31536000

HTTP/1.1 401 Unauthorized
content-type: application/json
docker-distribution-api-version: registry/2.0
www-authenticate: Bearer realm="https://auth.docker.io/token",service="registry.docker.io" <---------------------
date: Thu, 16 Jan 2025 10:51:12 GMT
content-length: 87
strict-transport-security: max-age=31536000

{"errors":[{"code":"UNAUTHORIZED","message":"authentication required","detail":null}]}
```

**Reply by invis-z** [reply]
In the case of a mirror, querying the mirror will return how it want to auth, most of the time this is not needed as `docker` actually doesn't handle mirror auth correctly. https://github.com/distribution/distribution/issues/1486#issuecomment-194390671
In this use case, since all the images are public accessible, if the mirror is set in daemon.json, it won't need to be authed so it will not need the token to access.

**Reply by szaimen** [reply]
I see. Thanks for the answers and explanation. 

So this would need a bigger refactoring to make it work...

I fear I don't have the time to work on this, so PRs are welcome!

**Comment by szaimen** [comment, marked as answer]
Hi, we publish the images nowadays on ghcr.io. does this improve the situation for you?
